### PR TITLE
Add banner announcing "Mozilla Monitor" rebrand

### PR DESCRIPTION
**Pull request by Vinnl**
# References: 
Jira: MNTOR-2522
Figma: https://www.figma.com/file/51jgudmkBoMMU9twT7UOKB/Monitor-Plus-Launch?type=design&node-id=310-6816&mode=design&t=PAVHr1PiURqF4TxK-0

# Description

(Based on https://github.com/mozilla/blurts-server/pull/4033 to avoid adding this to the wrong directory.)

# Screenshot (if applicable)

image

image

# How to test

First, enabled the `RebrandAnnouncement` feature flag.

Open the landing page. You should see the announcement. After you click the X, it should go away, and stay away after a refresh.

# Checklist (Definition of Done)
- [x] Localization strings (if needed) have been added.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [x] I've added or updated the relevant sections in readme and/or code comments
- [x] I've added a unit test to test for potential regressions of this bug.
- [ ] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [x] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [x] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: src/app/hooks/useLocalDismissal.ts
Comment (Vinnl) [inline src/app/hooks/useLocalDismissal.ts:1]: This hooks was based on [the equivalent hook in Relay](https://github.com/mozilla/fx-private-relay/blob/main/frontend/src/hooks/localDismissal.ts), except using (and working around :unamused:) the cookie library we were already using.

File: src/app/hooks/useHasRenderedClientSide.tsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { useEffect, useState } from "react";
+
+export function useHasRendedClientSide() {
```
Comment (pdehaan) [inline src/app/hooks/useHasRenderedClientSide.tsx:7]: Nit: "Rended" --> "Rendered" (plus the 2&times; other instances where this is imported and used above in `src/app/(proper_react)/(redesign)/(public)/RebrandAnnouncement.tsx`)
```suggestion
export function useHasRenderedClientSide() {
```

File: src/app/(proper_react)/(redesign)/(public)/RebrandAnnouncement.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import styles from "./RebrandAnnouncement.module.scss";
+import { CloseBtn } from "../../../components/server/Icons";
+import { useL10n } from "../../../hooks/l10n";
+import { useLocalDismissal } from "../../../hooks/useLocalDismissal";
+import { useHasRendedClientSide } from "../../../hooks/useHasRenderedClientSide";
+
+export const RebrandAnnouncement = () => {
+  const l10n = useL10n();
+  const dismissal = useLocalDismissal("mozilla-monitor-rebrand", {
+    duration: Number.MAX_SAFE_INTEGER,
```
Comment (pdehaan) [inline src/app/(proper_react)/(redesign)/(public)/RebrandAnnouncement.tsx:16]: What is the `duration` for here? Auto-closing?
Acccctually, per a doc comment below:
> ```js
> /** How long the dismissal should last, in seconds. Note that by default, the cookie will be deleted in 30 days. */
> ```

Where [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) says the value of `Number.MAX_SAFE_INTEGER` is ~9 quadrillion; and Google says 
that [9,007,199,254,740,991 seconds to years](https://www.google.com/search?q=9%2C007%2C199%2C254%2C740%2C991+s+to+years) is approximately 285_616_414 years (give or take a couple months).

File: src/app/hooks/useLocalDismissal.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+  const [isDismissed, setIsDismissed] = useState(
+    hasDismissedCookie(cookieValue, options.duration),
+  );
+
+  const dismiss = (dismissOptions?: DismissOptions) => {
+    const maxAgeInSeconds =
+      typeof options.duration === "number"
+        ? options.duration
+        : 100 * 365 * 24 * 60 * 60;
```
Comment (pdehaan) [inline src/app/hooks/useLocalDismissal.ts:46]: 100 years?

File: src/app/hooks/useHasRenderedClientSide.tsx (hunk 1 above)
Comment (Vinnl) [reply src/app/hooks/useHasRenderedClientSide.tsx]: Thank you! 690b5fccd

File: src/app/(proper_react)/(redesign)/(public)/RebrandAnnouncement.tsx (hunk 2 above)
Comment (Vinnl) [reply src/app/(proper_react)/(redesign)/(public)/RebrandAnnouncement.tsx:18]: Yep correct. I might've gone overboard with `MAX_SAFE_INTEGER` here, not showing it again for a 100 years would probably have been fine too :sweat_smile:

File: src/app/hooks/useLocalDismissal.ts (hunk 3 above)
Comment (Vinnl) [reply src/app/hooks/useLocalDismissal.ts:46]: "forever", basically. (There were a couple of instances on Relay where we show something again if it was dismissed more than `x` time ago. Maybe I should just remove it here until we find we actually need it.)

File: src/app/(proper_react)/(redesign)/(public)/LandingView.test.tsx [hunk 4]
```diff
@@ -164,6 +173,61 @@ describe("When Premium is not available", () => {
...
+
+  it("hides the banner announcing the rebrand to 'Mozilla Monitor' after clicking the dismiss button", async () => {
+    const ComposedDashboard = composeStory(LandingNonUs, Meta);
+    render(<ComposedDashboard />);
+
+    const rebrandAnnouncement = screen.getByText(
+      "New name, look and even more ways to",
+      { exact: false },
+    );
+    const dismissButton = within(rebrandAnnouncement.parentElement!).getByRole(
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/app/(proper_react)/(redesign)/(public)/LandingView.test.tsx:196]: I was actually hoping to not have to use it, but I didn't have a good way to target the complete banner. (I think it might need some accessible role, but I'm not sure which...)

File: src/app/(proper_react)/(redesign)/(public)/LandingView.test.tsx [hunk 5]
```diff
@@ -164,6 +173,61 @@ describe("When Premium is not available", () => {
...
+      "button",
+      { name: "OK" },
+    );
+
+    const user = userEvent.setup();
+    await user.click(dismissButton);
+
+    expect(mockedRecord).toHaveBeenCalledWith(
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/app/(proper_react)/(redesign)/(public)/LandingView.test.tsx:224]: This isn't calling `mockedRecord`, it's checking whether it _has been called_ by the code under test. In other words, I've replaced the `useTelemetry` function that the code calls by a test-specific function (see line 37 of this file), which returns a mock function (`mockedRecord`) to replace the function the code would usually call (`recordTelemetry`). The mock function doesn't actually do anything, other than keep track of how often it has been called, and with which parameters.

On this line, I'm checking that it has actually been called, with the parameters that I'd expect for tracking a button click (`"button"`, `"click"` and an object with the property `button_id` set to `"rebrand_announcement_dismiss"`).

So in other words, this test is not verifying that the banner disappears, but that the correct telemetry is sent.

File: src/app/(proper_react)/(redesign)/(public)/RebrandAnnouncement.module.scss [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  align-items: center;
+  justify-content: center;
+  gap: $spacing-xs;
+  padding: $spacing-lg;
+  background-color: $color-blue-20;
+  font: $text-body-lg;
+
+  @media screen and (max-width: $screen-md) {
+    // (button width) + (.rebrandAnnouncement gap) + (2 * inline button padding)
+    padding-inline-end: calc(24px + calc($spacing-xs + calc(2 * $spacing-sm)));
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/app/(proper_react)/(redesign)/(public)/RebrandAnnouncement.module.scss:17]: Ah! I suppose, I wasn't aware of what the `FalseDoorBanner` looked like. I felt like our banners got designed ad-hoc whenever one is needed, but if they are similar in designs, a general banner component indeed makes sense to me.

As for the specificity, what I wanted to do was add `$spacing-xs` padding in between the button and the text, but since the button is absolutely positioned, I couldn't just add that as a gap - I had to reproduce the button width and insert empty space on the right-hand side of the text. Since I had to calculate what that width would be anyway, I figured I'd might as well make it explicit in code to avoid a magic number and to make it easier to re-calculate if the button width is ever adjusted.

File: src/app/hooks/useHasRenderedClientSide.tsx [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+export function useHasRenderedClientSide() {
+  const [hasRenderedClientSide, setHasRenderedClientSide] = useState(false);
+
+  useEffect(() => {
+    setHasRenderedClientSide(true);
+  }, []);
+
+  return hasRenderedClientSide;
+}
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/app/hooks/useHasRenderedClientSide.tsx:15]: No, good point, I should've added that in the first place! Wrote a comment in 7ee9b9d3e; let me know if that makes sense or needs more elaboration.

File: src/app/hooks/useLocalDismissal.test.tsx [hunk 8]
```diff
@@ -0,0 +1,141 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { act, renderHook } from "@testing-library/react";
+import { useLocalDismissal } from "./useLocalDismissal";
+import { useCookies } from "react-cookie";
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/app/hooks/useLocalDismissal.test.tsx:7]: Indeed we'll generally assume that the library is working well (or rather, we'll have vetted that before we add it :sweat_smile: ). Thus, in these tests, we're not testing the library itself, but how our code interacts with it. On line 9 to 13 you can see that we mock the library, i.e. we replace it with our own implementation that doesn't do anything other than keeping track how often it is called, and with what arguments.

Then the tests ensure that we call that library the right way when we should be setting or removing cookies. But we'll leave whether those cookies _actually_ get set and removed up to the library itself, and assume that that's working well.

(Sometimes those assumptions break, or the way we're calling the library is actually wrong. In other words, the way we're integrating our code and the library is wrong. If we consider the risk of that to be high, we could write _integration tests_, or cover them in our end-to-end (Playwright) tests - but that's more of a hassle and [more likely to result in false positives](https://testing.googleblog.com/2017/04/where-do-our-flaky-tests-come-from.html), so we're not really doing that at this time.)

File: src/app/(proper_react)/(redesign)/(public)/LandingView.test.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vinnl) [reply src/app/(proper_react)/(redesign)/(public)/LandingView.test.tsx]: Maybe, I'm not sure. The general idea is to make our tests perform much the same actions as our users would do, and looking at the string "New name, look, etc" is obviously closer to that than looking at the source code and finding a test ID would be. But: looking for that string, then inspecting the DOM, looking for the parent element, and then finding a button in there of course _also_ isn't really what a user would do.

But with the response I got from the a11y team, I'm now simulating a user using landmark navigation to get to the `<aside>`, and then seeing the text in there: 9cac3055e. That might sort of be better?
